Modules/Scripted/VectorToScalarVolume/VectorToScalarVolume.py

```diff
@@ -26,82 +63,420 @@ def __init__(self, parent):
+    self.conversionMethodWidget.componentsComboBox.insertItem(self.conversionMethodWidget.EXTRACT_COMPONENT_NONE, "None")
+
+
+  def parameterNode(self):
+    return self._parameterNode
+
+  def setParameterNode(self, inputParameterNode):
+    if inputParameterNode == self._parameterNode:
+      return
+    if self._parameterNode is not None:
+      self.removeObserver(self._parameterNode, vtk.vtkCommand.ModifiedEvent, self.updateGuiFromMRML)
+    if inputParameterNode is not None:
+      self.addObserver(inputParameterNode, vtk.vtkCommand.ModifiedEvent, self.updateGuiFromMRML)
+    self._parameterNode = inputParameterNode
+
+  def inputVolumeNode(self):
+    return self.inputSelector.currentNode()
+
+  def setInputVolumeNode(self, node):
+    if isinstance(node, basestring):
+      node = getNode(node)
+    self.inputSelector.setCurrentNode(node)
+    self.conversionMethodWidget.setInputVolumeNode(node)
+
+  def outputVolumeNode(self):
+    return self.outputSelector.currentNode()
+
+  def setOutputVolumeNode(self, node):
+    if isinstance(node, basestring):
+      node = getNode(node)
+    self.outputSelector.setCurrentNode(node)
+
+  def updateGuiFromMRML(self, caller=None, event=None):
+    """
+    Query all the parameters in the parameterNode,
+    and update the GUI state accordingly if something has changed.
+    """
+    self.applyButton.enabled = all([self.inputVolumeNode(), self.outputVolumeNode(), self.parameterNode()])
+    if self.parameterNode is None:
+      return
+    self.setInputVolumeNode(self._parameterNode.GetParameter("InputVectorVolume"))
+    self.setOutputVolumeNode(self._parameterNode.GetParameter("OutputScalarVolume"))
+    self.conversionMethodWidget.setConversionMethod(self._parameterNode.GetParameter("ConversionMethod"))
+    self.conversionMethodWidget.setComponentToExtract(int(self._parameterNode.GetParameter("ComponentToExtract")))
+
+  def updateParameterNodeFromGui(self):
+
+    self.applyButton.enabled = all([self.inputVolumeNode(), self.outputVolumeNode(), self.parameterNode()])
+    if self._parameterNode is None:
+      return
+
+    with NodeModify(self._parameterNode):
+      self._parameterNode.SetParameter("InputVectorVolume", getNodeID(self.inputVolumeNode()))
+      self._parameterNode.SetParameter("OutputScalarVolume", getNodeID(self.outputVolumeNode()))
+      self._parameterNode.SetParameter("ConversionMethod", self.conversionMethodWidget.conversionMethod())
+      self._parameterNode.SetParameter("ComponentToExtract", str(self.conversionMethodWidget.componentToExtract()))
+
   def onApply(self):
-    inputVolume = self.inputSelector.currentNode()
-    outputVolume = self.outputSelector.currentNode()
-    # check for input data
-    if not (inputVolume and outputVolume):
-      slicer.util.errorDisplay('Input and output volumes are required for conversion', windowTitle='Luminance')
+
+    with ScopedQtPropertySetter(self.applyButton, {"enabled": False, "text": "Working..."}):
```

Comment:
> move it to slicer.util so that it can be used elsewhere.

That was indeed the plan. I suggest we do not use this call in this PR and introduced it only later.

Reply:
I just renamed it (it is too handy now to not use it!). If you want to move it to slicer.util afterwards with the good name, I can remove the local function here afterwards.